Rework User-Guide and Documentation (#1355)
viperproject/prusti: docs/user-guide/src/tour/pledges.md

With a pledge you can explain to Prusti how the original object gets affected by changes to the returned reference.
We will demonstrate it by implementing a function that gives you a mutable reference to the first element in the list:

## Implementing `peek_mut`

The `peek_mut` will return a mutable reference of type `T`, so the precondition of the list requires it to be non-empty.
As a first postcondition, we want to ensure that the `result` of `peek_mut` points to the first element of the list.

In the code, we need to get a mutable reference to the type inside the `Link = Option<Box<Node<T>>>`. This requires the use of the type `Option<&mut T>`, which is a structure containing a reference, so it is not yet supported by Prusti (see [limitations chapter](../capabilities/limitations.md)). To still be able to verify `peek_mut`, we mark it as `trusted` for now:

```rust,noplaypen
{{#rustdoc_include ../../../../prusti-tests/tests/verify/fail/user-guide/peek_mut_pledges.rs:peek_mut_code}}
```

Note that `peek_mut` cannot be `#[pure]`, since it returns a mutable reference.

## Writing a test for our specification

> Lets write a test to see if our specification works:

Reviewer: ```suggestion
Let's write a test to see if our specification works:
```

(did you mean to keep this part?)

- Create a list with two elements: [16, 8]
- Get a mutable reference to the first element (16)
- Change the first element to 5
- Check if the list still has the expected properties after `first` gets dropped
  - Is the length 2?
  - Is the first element 5?
  - Is the second element 8?

```rust,noplaypen
{{#rustdoc_include ../../../../prusti-tests/tests/verify/fail/user-guide/peek_mut_pledges.rs:test_peek_mut}}
```

But this fails, Prusti cannot verify any of our last three `prusti_assert` statements. This is where `pledges` come in. We have to tell Prusti how the `result` affects the original list. Without this, Prusti assumes that changes to the reference can change every property of the original list, so nothing can be known about it after the reference gets dropped.

## Writing the pledge

> The pledge gets written with an annotation like for `ensures` and `requires`, but with the keyword `after_expiry`.

Reviewer: ```suggestion
The pledge is written in an `after_expiry` attribute.
```

Inside we have all the conditions that hold after the returned reference gets dropped:

```rust,noplaypen
{{#rustdoc_include ../../../../prusti-tests/tests/verify/pass/user-guide/peek_mut_pledges.rs:pledge}}
```

We have 3 conditions here:
1. The list will have the same length afterwards
2. Any element of the list with index `1..list.len()` will not be changed
3. The element at the head of the list is the value that was assigned to the returned reference. This is denoted with the `before_expiry` function

With these 3 conditions, our test verifies successfully!

> ## Assert on expiry

Reviewer: Try to emphasise a bit more that the pledge precondition is an obligation that the *caller* needs to uphold.
